### PR TITLE
Add AnnotateQuery operator and FromSql extension method

### PR DESCRIPTION
**Pull request by mikary**
New PR for ```FromSql``` functionality. This is still a bit rough, in particular still using ```string.Format``` for parameters on the raw SQL string, but shows the general pattern of raw SQL through ```QueryAnnotation```s
@anpete @divega

### CODE REVIEW

File: src/EntityFramework.Core/DbSet`.cs [hunk 1]
```diff
@@ -275,5 +276,19 @@ IQueryProvider IQueryable.Provider
...
+        /// <summary>
+        ///     <para>
+        ///         Adds a <see cref="QueryAnnotation"/> to the query.
+        ///     </para>
+        ///     <para>
+        ///         This method is intended for use by extension methods that need to add information to
+        ///         the expression tree.
+        ///     </para>
+        /// </summary>
+        DbSet<TEntity> IDbSetExtender<TEntity>.AnnotateQuery(object annotation)
```
Comment (anpete) [inline src/EntityFramework.Core/DbSet`.cs]: Not sure about the name - perhaps something like IAnnotatableQuery?

File: src/EntityFramework.Core/Query/QueryAnnotationExtractor.cs [hunk 2]
```diff
@@ -30,8 +31,10 @@ public virtual ICollection<QueryAnnotation> ExtractQueryAnnotations([NotNull] Qu
         {
             foreach (var resultOperator
                 in queryModel.ResultOperators
-                    .Where(ro => ro is IncludeResultOperator
-                                 || ro is AsNoTrackingResultOperator)
+                    .Where(ro => ro.GetType()
```
Comment (anpete) [inline src/EntityFramework.Core/Query/QueryAnnotationExtractor.cs]: I don't think we need the attribute now.

File: src/EntityFramework.Relational/Properties/Strings.resx [hunk 3]
```diff
@@ -231,4 +231,7 @@
   <data name="DiscriminatorPropertyNotFound" xml:space="preserve">
     <value>Unable to set property '{property}' as a discriminator for entity type '{entityType}' because it is not a property of '{entityType}'.</value>
   </data>
+  <data name="MultipleFromSqlCalls" xml:space="preserve">
+    <value>Multiple calls to FromSql are not supported.</value>
```
Comment (anpete) [inline src/EntityFramework.Relational/Properties/Strings.resx]: We should try and prevent chaining of this. In particular, can we return IQueryable<> from FromSql?

File: src/EntityFramework.Relational/Query/Sql/DefaultSqlQueryGenerator.cs [hunk 4]
```diff
@@ -178,6 +179,32 @@ public virtual Expression VisitSelectExpression(SelectExpression selectExpressio
...
+
+            _sql.AppendLine("(");
+
+            using (_sql.Indent())
+            {
+                var sql = string.Format(rawSqlDerivedTableExpression.Sql, rawSqlDerivedTableExpression.Parameters);
+
+                using (var reader = new StringReader(sql))
+                {
+                    string line;
```
Comment (anpete) [inline src/EntityFramework.Relational/Query/Sql/DefaultSqlQueryGenerator.cs]: Add AppendLines to IndentedStringBuilder?

File: src/EntityFramework.Relational/RelationalDbSetExtensions.cs [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+using Microsoft.Data.Entity.Relational.Query.Annotations;
+using Microsoft.Data.Entity.Utilities;
+
+// ReSharper disable once CheckNamespace
+
+namespace Microsoft.Data.Entity
+{
+    public static class RelationalDbSetExtensions
+    {
+        public static DbSet<TEntity> FromSql<TEntity>([NotNull]this DbSet<TEntity> dbSet, [NotNull]string sql, [NotNull] params object[] parameters)
```
Comment (anpete) [inline src/EntityFramework.Relational/RelationalDbSetExtensions.cs]: Return IQueryable

**Comment by mikary** [comment]
Updated based on PR feedback. Cleaned up for adding basic changes, parameterized ```FromSql``` queries to be added in a later PR.

File: src/EntityFramework.Core/DbSet`.cs [hunk 6]
```diff
@@ -275,5 +276,19 @@ IQueryProvider IQueryable.Provider
...
+        /// <summary>
+        ///     <para>
+        ///         Adds a <see cref="QueryAnnotation"/> to the query.
+        ///     </para>
+        ///     <para>
+        ///         This method is intended for use by extension methods that need to add information to
+        ///         the expression tree.
+        ///     </para>
+        /// </summary>
+        IQueryable<TEntity> IAnnotatableQueryable<TEntity>.AnnotateQuery(object annotation)
```
Comment (anpete) [inline src/EntityFramework.Core/DbSet`.cs]: Does this compose? Can I add multiple of these to an ET?

File: src/EntityFramework.Core/Extensions/Internal/AnnotateQueryExtensions.cs [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+
+// ReSharper disable once CheckNamespace
+
+namespace System.Linq
+{
+    internal static class AnnotateQueryExtensions
+    {
+        internal static readonly MethodInfo AnnotateQueryMethodInfo
+            = typeof(AnnotateQueryExtensions)
+                .GetTypeInfo().GetDeclaredMethod("AnnotateQuery");
```
Comment (anpete) [inline src/EntityFramework.Core/Extensions/Internal/AnnotateQueryExtensions.cs]: Use nameof

File: src/EntityFramework.Core/Query/QueryCompilationContext.cs [hunk 8]
```diff
@@ -46,6 +47,8 @@ public abstract class QueryCompilationContext
 
         public virtual IEntityKeyFactorySource EntityKeyFactorySource { get; }
 
+        public virtual ICollection<QueryAnnotation> QueryAnnotations { get; [param: NotNull] set; }
```
Comment (anpete) [inline src/EntityFramework.Core/Query/QueryCompilationContext.cs]: Add a Check.NotNull to the setter

File: src/EntityFramework.Core/Query/ResultOperators/AnnotateQueryExpressionNode.cs [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+    public class AnnotateQueryExpressionNode : ResultOperatorExpressionNodeBase
+    {
+        public static readonly MethodInfo[] SupportedMethods = { AnnotateQueryExtensions.AnnotateQueryMethodInfo };
+
+        private readonly ConstantExpression _annotationExpression;
+
+        public AnnotateQueryExpressionNode(
+            MethodCallExpressionParseInfo parseInfo,
+            [NotNull] ConstantExpression annotationExpression)
+            : base(parseInfo, null, null)
```
Comment (anpete) [inline src/EntityFramework.Core/Query/ResultOperators/AnnotateQueryExpressionNode.cs]: Null check inline and attribute.

File: src/EntityFramework.Core/Query/ResultOperators/AnnotateQueryExpressionNode.cs [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+        {
+            Check.NotNull(annotationExpression, nameof(annotationExpression));
+
+            _annotationExpression = annotationExpression;
+        }
+
+        protected override ResultOperatorBase CreateResultOperator(ClauseGenerationContext clauseGenerationContext)
+        {
+            return new AnnotateQueryResultOperator(_annotationExpression);
```
Comment (anpete) [inline src/EntityFramework.Core/Query/ResultOperators/AnnotateQueryExpressionNode.cs]: Null check and attribute.

File: src/EntityFramework.Core/Query/ResultOperators/AnnotateQueryResultOperator.cs [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+            return new AnnotateQueryResultOperator(_annotationExpression);
+        }
+
+        public override void TransformExpressions([NotNull] Func<Expression, Expression> transformation)
+        {
+        }
+
+        public override StreamedSequence ExecuteInMemory<T>([NotNull] StreamedSequence input)
+        {
+            Check.NotNull(input, nameof(input));
```
Comment (anpete) [inline src/EntityFramework.Core/Query/ResultOperators/AnnotateQueryResultOperator.cs]: Can skip the check here

File: src/EntityFramework.Relational/Query/Annotations/FromSqlAnnotation.cs [hunk 12]
```diff
@@ -0,0 +1,25 @@
...
+    public class FromSqlAnnotation
+    {
+        public FromSqlAnnotation([NotNull] string sql)
+        {
+            Check.NotEmpty(sql, nameof(sql));
+
+            Sql = sql;
+        }
+
+        public virtual string Sql { get; private set; }
```
Comment (anpete) [inline src/EntityFramework.Relational/Query/Annotations/FromSqlAnnotation.cs]: Can remove private set

File: src/EntityFramework.Relational/Query/ExpressionTreeVisitors/RelationalEntityQueryableExpressionTreeVisitor.cs [hunk 13]
```diff
@@ -78,15 +81,29 @@ protected override Expression VisitEntityQueryable(Type elementType)
...
-                .AddTable(
+            var alias = _querySource.ItemName.StartsWith("<generated>_")
+                ? tableName.First().ToString().ToLower()
+                : _querySource.ItemName;
+
+            var fromSqlAnnotation = GetAnnotations<FromSqlAnnotation>(_querySource).SingleOrDefault();
+
+            if(fromSqlAnnotation != null)
+            {
+                selectExpression.AddTable(
```
Comment (anpete) [inline src/EntityFramework.Relational/Query/ExpressionTreeVisitors/RelationalEntityQueryableExpressionTreeVisitor.cs]: Optional: Can write this with a single AddTable call with an embedded conditional expr.

File: src/EntityFramework.Relational/RelationalDbSetExtensions.cs [hunk 14]
```diff
@@ -0,0 +1,25 @@
...
+{
+    public static class RelationalDbSetExtensions
+    {
+        public static IQueryable<TEntity> FromSql<TEntity>([NotNull]this DbSet<TEntity> dbSet, [NotNull]string sql)
+            where TEntity : class
+        {
+            Check.NotNull(dbSet, nameof(dbSet));
+            Check.NotEmpty(sql, nameof(sql));
+
+            return (dbSet as IAnnotatableQueryable<TEntity>).AnnotateQuery(new FromSqlAnnotation(sql));
```
Comment (anpete) [inline src/EntityFramework.Relational/RelationalDbSetExtensions.cs]: Use a Cast here.

File: src/EntityFramework.Core/Extensions/Internal/AnnotateQueryExtensions.cs [hunk 15]
```diff
@@ -0,0 +1,30 @@
...
+
+        internal static IQueryable<TEntity> AnnotateQuery<TEntity>([NotNull] this IQueryable<TEntity> source, [NotNull] object annotation) where TEntity : class
```
Comment (anpete) [inline src/EntityFramework.Core/Extensions/Internal/AnnotateQueryExtensions.cs]: Make this public but move it to the Query namespace.

**Comment by anpete** [comment]
Looks great! :shipit: after making the changes above.

**Pull request merged**